pkg/webhooks/tensorflow/tfjob_webhook.go

```diff
@@ -0,0 +1,121 @@
+	tfReplicaSpecPath = specPath.Child("tfReplicaSpecs")
+)
+
+type Webhook struct{}
+
+func SetupWebhook(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(&trainingoperator.TFJob{}).
+		WithValidator(&Webhook{}).
+		Complete()
+}
+
+// +kubebuilder:webhook:path=/validate-kubeflow-org-v1-tfjob,mutating=false,failurePolicy=fail,sideEffects=None,groups=kubeflow.org,resources=tfjobs,verbs=create;update,versions=v1,name=validator.tfjob.training-operator.kubeflow.org,admissionReviewVersions=v1
+
+var _ webhook.CustomValidator = &Webhook{}
+
+func (w *Webhook) ValidateCreate(ctx context.Context, obj runtime.Object) (admission.Warnings, error) {
+	job := obj.(*trainingoperator.TFJob)
+	log := ctrl.LoggerFrom(ctx).WithName("tfjob-webhook")
+	log.V(5).Info("Validating create", "TFJob", klog.KObj(job))
+	return nil, validateTFJob(job).ToAggregate()
+}
+
+func (w *Webhook) ValidateUpdate(ctx context.Context, _, newObj runtime.Object) (admission.Warnings, error) {
+	job := newObj.(*trainingoperator.TFJob)
+	log := ctrl.LoggerFrom(ctx).WithName("tfjob-webhook")
+	log.V(5).Info("Validating update", "NewTFJob", klog.KObj(job))
+	return nil, validateTFJob(job).ToAggregate()
+}
+
+func (w *Webhook) ValidateDelete(context.Context, runtime.Object) (admission.Warnings, error) {
+	return nil, nil
+}
+
+func validateTFJob(job *trainingoperator.TFJob) field.ErrorList {
+	var allErrs field.ErrorList
+	if errors := apimachineryvalidation.NameIsDNS1035Label(job.Name, false); len(errors) != 0 {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("metadata").Child("name"), job.Name, fmt.Sprintf("should match: %v", strings.Join(errors, ","))))
+	}
+	allErrs = append(allErrs, validateSpec(job.Spec)...)
+	return allErrs
+}
+
+func validateSpec(spec trainingoperator.TFJobSpec) field.ErrorList {
+	return validateTFReplicaSpecs(spec.TFReplicaSpecs)
+}
+
+func validateTFReplicaSpecs(rSpecs map[trainingoperator.ReplicaType]*trainingoperator.ReplicaSpec) field.ErrorList {
+	var allErrs field.ErrorList
+
+	if rSpecs == nil {
+		allErrs = append(allErrs, field.Required(tfReplicaSpecPath, "must be required"))
+	}
+
+	chiefOrMaster := 0
+	for rType, rSpec := range rSpecs {
+		rolePath := tfReplicaSpecPath.Key(string(rType))
+		containerPath := rolePath.Child("template").Child("spec").Child("containers")
+
+		if rSpec == nil || len(rSpec.Template.Spec.Containers) == 0 {
+			allErrs = append(allErrs, field.Required(containerPath, "must be specified"))
+		}
+		if trainingoperator.IsChiefOrMaster(rType) {
+			chiefOrMaster++
+		}
+		// Make sure the image is defined in the container.
+		defaultContainerPresent := false
+		for idx, container := range rSpec.Template.Spec.Containers {
+			if container.Image == "" {
+				allErrs = append(allErrs, field.Required(containerPath.Index(idx).Child("image"), "must be required"))
+			}
+			if container.Name == trainingoperator.TFJobDefaultContainerName {
+				defaultContainerPresent = true
+			}
+		}
```

Comment:
Similar to above comment maybe that could be moved to the helper function in which we pass container name for the check.